bin/p2-rm/main.go

```diff
@@ -40,37 +67,123 @@ func main() {
 		*nodeName = hostname
 	}
 
-	// If the node is managed by a controller, let the user know. Changes to the
-	// controller might be needed.
-	labelID := path.Join(*nodeName, *podName)
-	podLabels, err := labeler.GetLabels(labels.POD, labelID)
+	rm := NewP2RM(kp.NewConsulClient(opts), *podName, *nodeName)
+
+	podIsManagedByRC, rcID, err := rm.checkForManagingReplicationController()
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "unable to check node for labels: %v\n", err)
-		os.Exit(1)
+		os.Exit(2)
 	}
-	if podLabels.Labels.Has(rc.RCIDLabel) {
+
+	if !podIsManagedByRC {
+		err = rm.deletePod()
+		if err != nil {
+			os.Exit(2)
+		}
+	}
+
+	if podIsManagedByRC && !*deallocation {
 		fmt.Fprintf(
 			os.Stderr,
-			"warning: %s is managed by replication controller: %v\n"+
-				"         To reduce the controller's capacity, use \"p2-rctl set-replicas\"\n"+
-				"continue [y|n]? ",
-			labelID,
-			podLabels.Labels.Get(rc.RCIDLabel),
-		)
-		input, err := bufio.NewReader(os.Stdin).ReadString('\n')
-		answer := (strings.ToLower(strings.TrimSpace(input)) + "n")[0]
-		if err != nil || answer != 'y' {
-			fmt.Fprintf(os.Stderr, "aborting\n")
+			"error: %s is managed by replication controller: %v\n"+
+				"It's possible you meant you deallocate this pod on this node. If so, please confirm your intention with --deallocate\n", rcID)
+		os.Exit(2)
+	}
+
+	if podIsManagedByRC && *deallocation {
+		err = rm.decrementDesiredCount(rcID)
+		if err != nil {
+			fmt.Fprintf(os.Stderr,
+				"Encountered error deallocating from the RC %s. You may attempt this command again or use `p2-rctl` to cleanup manually.\n%v",
+				rcID,
+				err)
+		}
+	}
+
+	fmt.Printf("%s: successfully removed %s\n", rm.NodeName, rm.PodName)
+}
+
+func sessionName(rcID fields.ID) string {
+	currentUser, err := user.Current()
+	username := "unknown"
+	if err == nil {
+		username = currentUser.Username
+	}
+
+	return fmt.Sprintf("p2-rm:user:%s:rcID:%s", username, rcID)
+}
+
+func (rm *P2RM) decrementDesiredCount(id fields.ID) error {
+	session, renewalErrChan, err := rm.Store.NewSession(sessionName(id), nil)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Unable to get consul session. Halting \n%v\n", err)
+		os.Exit(1)
+	}
+
+	go func(errChan chan error) {
+		select {
+		case <-errChan:
+			fmt.Fprintf(os.Stderr, "Got fatal renewal error!\n%v\n", err)
 			os.Exit(1)
+		default:
 		}
+	}(renewalErrChan)
+
+	rcLock, err := rm.RCStore.LockForMutation(id, session)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Unable to lock RC for mutation. Halting.\n %v", err)
+		os.Exit(1)
```

Comment:
and it would be very improbable 1) for this operation to last longer than a lock session (15 seconds i believe by default and 2) for the renewal to fail even if it did take that long

so in practice this should all be moot :)